pg_search/src/postgres/merge.rs

```diff
@@ -0,0 +1,286 @@
+            .map(|entry| entry.byte_size())
+            .sum::<u64>();
+        let target_segment_count = index_options.target_segment_count();
+        let target_byte_size = index_byte_size / target_segment_count as u64;
+
+        // clamp the highest layer size to be less than the following:
+        //
+        // `index_byte_size` / `target_segment_count`
+        //
+        // i.e. how big should each segment be if we want to have exactly `target_segment_count` segments?
+        //
+        // for instance, imagine:
+        // - layer sizes: [1mb, 10mb, 100mb]
+        // - index size: 200mb
+        // - target segment count: 10
+        //
+        // then our recomputed layer sizes would be:
+        // - [1mb, 10mb]
+        //
+        // why? the 100mb layer gets excluded because the target segment size is 20mb
+        pgrx::debug1!("target_byte_size for merge: {target_byte_size}");
+
+        let mut foreground_layer_sizes = index_options.foreground_layer_sizes();
+        foreground_layer_sizes.retain(|&layer_size| layer_size < target_byte_size);
+        pgrx::debug1!("adjusted foreground_layer_sizes: {foreground_layer_sizes:?}");
+
+        let mut background_layer_sizes = index_options.background_layer_sizes();
+        background_layer_sizes.retain(|&layer_size| layer_size < target_byte_size);
+        pgrx::debug1!("adjusted background_layer_sizes {background_layer_sizes:?}");
+
+        Self {
+            foreground_layer_sizes,
+            background_layer_sizes,
+        }
+    }
+}
+impl IndexLayerSizes {
+    fn foreground(&self) -> Vec<u64> {
+        self.foreground_layer_sizes.clone()
+    }
+
+    fn background(&self) -> Vec<u64> {
+        self.background_layer_sizes.clone()
+    }
+}
+
+/// Kick off a merge of the index, if needed.
+///
+/// First merge into the smaller layers in the foreground,
+/// then launch a background worker to merge down the larger layers.
+pub unsafe fn do_merge(index: &PgSearchRelation) -> anyhow::Result<()> {
+    let merger = SearchIndexMerger::open(MvccSatisfies::Mergeable.directory(index))?;
+    let layer_sizes = IndexLayerSizes::from(index);
+    let foreground_layers = layer_sizes.foreground();
+    let background_layers = layer_sizes.background();
+
+    let metadata = MetaPage::open(index);
+    let merge_lock = metadata.acquire_merge_lock();
+
+    let needs_background_merge = !background_layers.is_empty() && {
+        let mut background_merge_policy = LayeredMergePolicy::new(background_layers.clone());
+        background_merge_policy.set_mergeable_segment_entries(&metadata, &merge_lock, &merger);
+        let merge_candidates = background_merge_policy.simulate();
+        !merge_candidates.is_empty()
+    };
+
+    // first merge down the foreground layers
+    if !foreground_layers.is_empty() {
+        let foreground_merge_policy = LayeredMergePolicy::new(foreground_layers);
+        unsafe { foreground_merge_policy.merge_index(index, merge_lock, false) };
+    }
```

Comment:
this needs to drop the merge_lock if we don't do a foreground merge.  Otherwise we'll potentially try to launch the background worker while holding it.

```suggestion
    } else {
        // we no longer need to hold the [`MergeLock`] as we're not merging in the foreground
        drop(merge_lock);
    }
```